plonky3/src/stark.rs

```diff
@@ -169,46 +196,44 @@ where
         witness: &[(String, Vec<T>)],
         witgen_callback: WitgenCallback<T>,
     ) -> Result<Vec<u8>, String> {
-        let stage_0_trace =
-            generate_matrix(witness.iter().map(|(name, value)| (name, value.as_ref())));
+        // here we need to clone the witness because the callback will modify it
+        let witness = &mut witness.to_vec();
```

Comment:
chatgpt:

> Why does this work?
> The key lies in Rust's temporary lifetime extension. When a temporary is created and immediately borrowed, Rust extends the lifetime of that temporary until the end of the enclosing scope. So, in this case:

> Rust sees that you are creating a mutable reference (&mut) to a temporary value (witness.to_vec()).
> It extends the lifetime of the temporary Vec so that it lives as long as the reference (i.e., the temporary Vec will not be dropped until witness is no longer used).
> Effectively, what happens is that the Vec will remain valid for as long as the mutable reference &mut witness is in scope.

So in this particular case, Rust introduces an intermediate variable